walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -18,99 +18,154 @@
+    Comparator<Event> EventStartTimeComparator
         = Comparator.comparing(Event::getWhen, (e1, e2) -> {
             return Long.compare(e1.start(), e2.start());
         });
 
-    Collections.sort(allEvents, EventStartComparator);
+    Collections.sort(allEvents, EventStartTimeComparator);
+    Collections.sort(eventsWithoutOptionalAttendees, EventStartTimeComparator);
+
+    // Find valid gaps if we considered everyone.
+    Collection<TimeRange> gapsConsideringEveryone = findGaps(allEvents, meetingTime, numAttendees + numOptionalAttendees);
+
+    if (gapsConsideringEveryone.size() == 0) {
+        // If it's impossible to find any gaps when considering everyone,
+        // find valid gaps without considering optional attendees and return that instead.
+        Collection<TimeRange> gapsWithoutOptional = findGaps(eventsWithoutOptionalAttendees, meetingTime, numAttendees);
+        return gapsWithoutOptional;
+    } else {
+        // If there is at least one valid gap that considers everyone, that is better than
+        // having more valid gaps that exclude optional attendees, so we default to returning gaps
+        // that consider everyone.
+        return gapsConsideringEveryone;
+    }
+  }
+
+  /** Discard an event if none of the attendees who want to book a meeting are going to it. */
+  private List<Event> removePeopleNotAttending(List<Event> allEvents, Collection<String> allAttendees, Collection<String> allOptionalAttendees) {
+    Iterator<Event> allEventsIter = allEvents.iterator();
+
+    while (allEventsIter.hasNext()) {
+        Event e = allEventsIter.next();
+        if (eventAttendeeNotMeeting(e, allAttendees, allOptionalAttendees)) {
+            allEventsIter.remove();
+        }
+    }
+    return allEvents;
+  }
+
+  /** Return `true` if the person attending the event isn't one of the people who wants to book a meeting. */
+  private boolean eventAttendeeNotMeeting(Event e, Collection<String> allAttendees, Collection<String> allOptionalAttendees) {
+    Set<String> attendees = e.getAttendees();
     
-    // Track latest end time of any event.
-    // Since eventTimes was sorted by start, the event with the latest end time
-    // might come before the last event in the List (e.g. if we had [5, 20]
-    // followed by [10, 15], we want to use [20, END_OF_DAY] for the final gap,
-    // not [15, END_OF_DAY]).
+    for (String s : attendees) {
+        if ( !( allAttendees.contains(s) || allOptionalAttendees.contains(s)) ) {
+            return true;
+        }
+    }
+    return false;
+  }
+
+  /** Generate list of events that excludes optional attendees. */
+  private List<Event> getEventsWithoutOptionalAttendees(List<Event> allEvents, Collection<String> allOptionalAttendees) {
+    List<Event> eventsWithoutOptionalAttendees = new ArrayList(allEvents);
+    Iterator<Event> optionalEventsIter = eventsWithoutOptionalAttendees.iterator();
+    
+    while(optionalEventsIter.hasNext()) {
```

Comment:
Nice use of iterators.
This is especially good for iterating over general lists, since it will be more efficient than an indexed lookup when the underlying implementation isn't an ArrayList.